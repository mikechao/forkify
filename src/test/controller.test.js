/**
 * @jest-environment jsdom
 */
import { async } from 'regenerator-runtime';
import * as testHelper from './testHelper';

const html = testHelper.indexHTMLContent().toString();
process.env['NODE_DEV'] = 'TEST-112233';

async function getController() {
  let controllerM;
  await jest.isolateModulesAsync(async () => {
    controllerM = await import('../js/controller');
  });
  return controllerM.default;
}
describe('controller test controlRecipes', () => {
  beforeEach(() => {
    document.documentElement.innerHTML = html;
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  it('should not call any views', async () => {
    const controller = await getController();
    await controller.controlRecipes();
    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it('should render correctly', async () => {
    window.location.hash = '#' + testHelper.getNonBookmarkedRecipeId();
    const mockSearchRecipesResults = testHelper.searchRecipesResults();
    const mockGetNutritionWidget = jest
      .fn()
      .mockImplementation(testHelper.nutritionWdigetContent());
    const mockBookmarks = testHelper.bookmarkedRecipes();
    const mockRecipe = testHelper.getRecipe();
    jest.mock('../js/model', () => {
      const originalModule = jest.requireActual('../js/model');
      return {
        __esModule: true,
        ...originalModule,
        loadRecipe: jest.fn().mockImplementation(() => Promise.resolve('test')),
        getNutritionWidget: mockGetNutritionWidget,
        state: {
          bookmarks: mockBookmarks,
          recipe: mockRecipe,
          search: {
            results: mockSearchRecipesResults,
            page: 23,
          },
        },
      };
    });
    const controller = await getController();
    await controller.controlRecipes();
    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });
});
