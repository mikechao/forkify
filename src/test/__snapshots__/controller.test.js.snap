// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller test controlRecipes should not call any views 1`] = `
"<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <link href="https://fonts.googleapis.com/css?family=Nunito+Sans:400,600,700" rel="stylesheet">
    <link rel="shortcut icon" href="src/img/favicon.png" type="image/x-icon">

    <link rel="stylesheet" href="src/sass/main.scss">
    <script type="module" defer="" src="src/js/controller.js"></script>

    <title>forkify // Search over 1,000,000 recipes</title>
  </head>

  <body>
    <div class="container">
      <header class="header">
        <a href="/">
          <img src="src/img/logo.png" alt="Logo" class="header__logo">
        </a>
        <form class="search">
          <input type="text" class="search__field" placeholder="Search over 1,000,000 recipes...">
          <button class="btn search__btn">
            <svg class="search__icon">
              <use href="src/img/icons.svg#icon-search"></use>
            </svg>
            <span>Search</span>
          </button>
        </form>

        <nav class="nav">
          <ul class="nav__list">
            <li class="nav__item">
              <button class="nav__btn nav__btn--add-recipe">
                <svg class="nav__icon">
                  <use href="src/img/icons.svg#icon-edit"></use>
                </svg>
                <span>Add recipe</span>
              </button>
            </li>
            <li class="nav__item">
              <button class="nav__btn nav__btn--bookmarks">
                <svg class="nav__icon">
                  <use href="src/img/icons.svg#icon-bookmark"></use>
                </svg>
                <span>Bookmarks</span>
              </button>
              <div class="bookmarks">
                <ul class="bookmarks__list">
                  <div class="message">
                    <div>
                      <svg>
                        <use href="src/img/icons.svg#icon-smile"></use>
                      </svg>
                    </div>
                    <p>
                      No bookmarks yet. Find a nice recipe and bookmark it :)
                    </p>
                  </div>
                </ul>
              </div>
            </li>
          </ul>
        </nav>
      </header>

      <div class="search-results">
        <ul class="results"></ul>

        <div class="pagination"></div>

        <p class="copyright">
          © Copyright by
          <a class="twitter-link" target="_blank" href="https://twitter.com/jonasschmedtman">Jonas Schmedtmann</a>. Use for learning or your portfolio. Don't use to teach. Don't claim
          as your own.
        </p>
      </div>

      <div class="recipe">
        <div class="message">
          <div>
            <svg>
              <use href="src/img/icons.svg#icon-smile"></use>
            </svg>
          </div>
          <p>Start by searching for a recipe or an ingredient. Have fun!</p>
        </div>
      </div>
    </div>

    <div class="del-modal hidden">
      <button class="btn--close-del-modal">×</button>
      <div class="del-modal__content">
        <h3 class="del-modal__heading delContent">
          Are you sure you want to delete this recipe?
        </h3>
        <div class="del-modal__preview delContent"></div>
        <div class="del-modal__buttons delContent">
          <button class="btn del-modal__cancel-btn" type="button">
            <span>Cancel</span>
          </button>
          <button class="btn del-modal__delete-btn" type="button">
            <span>Delete</span>
          </button>
        </div>
      </div>
    </div>

    <div class="overlay hidden"></div>
    <div class="add-recipe-window hidden">
      <button class="btn--close-modal">×</button>
      <form class="upload">
        <button class="btn--inline btn--add-ing" type="button">
          <svg>
            <use href="src/img/icons.svg#icon-plus-circle"></use>
          </svg>
          <span>Add</span>
        </button>
        <button class="btn--inline btn--remove-ing" type="button">
          <svg>
            <use href="src/img/icons.svg#icon-minus-circle"></use>
          </svg>
          <span>Remove</span>
        </button>
        <div class="upload__column data__column">
          <h3 class="upload__heading">Recipe data</h3>
          <label>Title</label>
          <input value="TEST23" required="" name="title" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="titleError" class="hidden">Please enter a value for Title</span>
          <label>URL</label>
          <input value="TEST23" required="" name="sourceUrl" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="sourceUrlError" class="hidden">Please enter a value for URL</span>
          <label>Image URL</label>
          <input value="TEST23" required="" name="image" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="imageError" class="hidden">Please enter a value for Image URL</span>
          <label>Publisher</label>
          <input value="TEST23" required="" name="publisher" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="publisherError" class="hidden">Please enter a value for Publisher</span>
          <label>Prep time</label>
          <input value="23" required="" name="cookingTime" type="number" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="cookingTimeError" class="hidden">Please enter a value for Prep time</span>
          <label>Servings</label>
          <input value="23" required="" name="servings" type="number" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="servingsError" class="hidden">Please enter a value for Servings</span>
        </div>

        <div class="upload__column ing__column">
          <h3 class="upload__heading">Ingredients</h3>
          <label for="ingredient-1">Ingredient 1</label>
          <input value="1,,Avocado" type="text" required="" name="ingredient-1" placeholder="Format: 'Quantity,Unit,Description'" data-validate="isIngredient required">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-1Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-2">Ingredient 2</label>
          <input value=",,salt" type="text" name="ingredient-2" placeholder="Format: 'Quantity,Unit,Description'" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-2Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-3" class="hidden">Ingredient 3</label>
          <input type="text" name="ingredient-3" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-3Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-4" class="hidden">Ingredient 4</label>
          <input type="text" name="ingredient-4" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-4Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-5" class="hidden">Ingredient 5</label>
          <input type="text" name="ingredient-5" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-5Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-6" class="hidden">Ingredient 6</label>
          <input type="text" name="ingredient-6" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-6Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
        </div>

        <button class="btn upload__btn">
          <svg>
            <use href="src/img/icons.svg#icon-upload-cloud"></use>
          </svg>
          <span>Upload</span>
        </button>
      </form>
    </div>
  

</body>"
`;

exports[`controller test controlRecipes should render correctly 1`] = `
"<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <link href="https://fonts.googleapis.com/css?family=Nunito+Sans:400,600,700" rel="stylesheet">
    <link rel="shortcut icon" href="src/img/favicon.png" type="image/x-icon">

    <link rel="stylesheet" href="src/sass/main.scss">
    <script type="module" defer="" src="src/js/controller.js"></script>

    <title>forkify // Search over 1,000,000 recipes</title>
  </head>

  <body>
    <div class="container">
      <header class="header">
        <a href="/">
          <img src="src/img/logo.png" alt="Logo" class="header__logo">
        </a>
        <form class="search">
          <input type="text" class="search__field" placeholder="Search over 1,000,000 recipes...">
          <button class="btn search__btn">
            <svg class="search__icon">
              <use href="src/img/icons.svg#icon-search"></use>
            </svg>
            <span>Search</span>
          </button>
        </form>

        <nav class="nav">
          <ul class="nav__list">
            <li class="nav__item">
              <button class="nav__btn nav__btn--add-recipe">
                <svg class="nav__icon">
                  <use href="src/img/icons.svg#icon-edit"></use>
                </svg>
                <span>Add recipe</span>
              </button>
            </li>
            <li class="nav__item">
              <button class="nav__btn nav__btn--bookmarks">
                <svg class="nav__icon">
                  <use href="src/img/icons.svg#icon-bookmark"></use>
                </svg>
                <span>Bookmarks</span>
              </button>
              <div class="bookmarks">
                <ul class="bookmarks__list">
                  <div class="message">
                    <div>
                      <svg>
                        <use href="src/img/icons.svg#icon-smile"></use>
                      </svg>
                    </div>
                    <p>
                      No bookmarks yet. Find a nice recipe and bookmark it :)
                    </p>
                  </div>
                </ul>
              </div>
            </li>
          </ul>
        </nav>
      </header>

      <div class="search-results">
        <ul class="results"></ul>

        <div class="pagination"></div>

        <p class="copyright">
          © Copyright by
          <a class="twitter-link" target="_blank" href="https://twitter.com/jonasschmedtman">Jonas Schmedtmann</a>. Use for learning or your portfolio. Don't use to teach. Don't claim
          as your own.
        </p>
      </div>

      <div class="recipe">
        <div class="error">
            <div>
              <svg>
                <use href="[object Object]#icon-alert-triangle"></use>
              </svg>
            </div>
            <p>We could not find that recipe. Please try another one!</p>
        </div>
    </div>
    </div>

    <div class="del-modal hidden">
      <button class="btn--close-del-modal">×</button>
      <div class="del-modal__content">
        <h3 class="del-modal__heading delContent">
          Are you sure you want to delete this recipe?
        </h3>
        <div class="del-modal__preview delContent"></div>
        <div class="del-modal__buttons delContent">
          <button class="btn del-modal__cancel-btn" type="button">
            <span>Cancel</span>
          </button>
          <button class="btn del-modal__delete-btn" type="button">
            <span>Delete</span>
          </button>
        </div>
      </div>
    </div>

    <div class="overlay hidden"></div>
    <div class="add-recipe-window hidden">
      <button class="btn--close-modal">×</button>
      <form class="upload">
        <button class="btn--inline btn--add-ing" type="button">
          <svg>
            <use href="src/img/icons.svg#icon-plus-circle"></use>
          </svg>
          <span>Add</span>
        </button>
        <button class="btn--inline btn--remove-ing" type="button">
          <svg>
            <use href="src/img/icons.svg#icon-minus-circle"></use>
          </svg>
          <span>Remove</span>
        </button>
        <div class="upload__column data__column">
          <h3 class="upload__heading">Recipe data</h3>
          <label>Title</label>
          <input value="TEST23" required="" name="title" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="titleError" class="hidden">Please enter a value for Title</span>
          <label>URL</label>
          <input value="TEST23" required="" name="sourceUrl" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="sourceUrlError" class="hidden">Please enter a value for URL</span>
          <label>Image URL</label>
          <input value="TEST23" required="" name="image" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="imageError" class="hidden">Please enter a value for Image URL</span>
          <label>Publisher</label>
          <input value="TEST23" required="" name="publisher" type="text" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="publisherError" class="hidden">Please enter a value for Publisher</span>
          <label>Prep time</label>
          <input value="23" required="" name="cookingTime" type="number" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="cookingTimeError" class="hidden">Please enter a value for Prep time</span>
          <label>Servings</label>
          <input value="23" required="" name="servings" type="number" data-validate="required">
          <span id="spacer" class="hidden"></span>
          <span id="servingsError" class="hidden">Please enter a value for Servings</span>
        </div>

        <div class="upload__column ing__column">
          <h3 class="upload__heading">Ingredients</h3>
          <label for="ingredient-1">Ingredient 1</label>
          <input value="1,,Avocado" type="text" required="" name="ingredient-1" placeholder="Format: 'Quantity,Unit,Description'" data-validate="isIngredient required">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-1Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-2">Ingredient 2</label>
          <input value=",,salt" type="text" name="ingredient-2" placeholder="Format: 'Quantity,Unit,Description'" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-2Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-3" class="hidden">Ingredient 3</label>
          <input type="text" name="ingredient-3" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-3Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-4" class="hidden">Ingredient 4</label>
          <input type="text" name="ingredient-4" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-4Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-5" class="hidden">Ingredient 5</label>
          <input type="text" name="ingredient-5" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-5Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
          <label for="ingredient-6" class="hidden">Ingredient 6</label>
          <input type="text" name="ingredient-6" placeholder="Format: 'Quantity,Unit,Description'" class="hidden" data-validate="isIngredient">
          <span id="spacer" class="hidden"></span>
          <span id="ingredient-6Error" class="hidden">Please use the Format: 'Quantity,Unit,Description'</span>
        </div>

        <button class="btn upload__btn">
          <svg>
            <use href="src/img/icons.svg#icon-upload-cloud"></use>
          </svg>
          <span>Upload</span>
        </button>
      </form>
    </div>
  

</body>"
`;
